prow/job/sippy.py

```diff
@@ -0,0 +1,230 @@
+        self._params.update(sampleRelease=release)
+        return self
+
+    def sample_starttime(self, starttime):
+        self._params.update(sampleStartTime=starttime)
+        return self
+
+    def sample_endtime(self, endtime):
+        self._params.update(sampleEndTime=endtime)
+        return self
+
+    def confidence(self, confidence="95"):
+        self._params.update(confidence=confidence)
+        return self
+
+    def ignore_disruption(self, ignore="true"):
+        self._params.update(ignoreDisruption=ignore)
+        return self
+
+    def ignore_missing(self, ignore="false"):
+        self._params.update(ignoreMissing=ignore)
+        return self
+
+    def exclude_arches(self, arches="arm64,heterogeneous,ppc64le,s390x"):
+        self._params.update(excludeArches=arches)
+        return self
+
+    def exclude_clouds(self, clouds="openstack,ibmcloud,libvirt,ovirt,unknown"):
+        self._params.update(excludeClouds=clouds)
+        return self
+
+    def exclude_variants(self, variants="hypershift,osd,microshift,techpreview,single-node,assisted,compact"):
+        self._params.update(excludeVariants=variants)
+        return self
+
+    def group_by(self, group_by="cloud,arch,network"):
+        self._params.update(groupBy=group_by)
+        return self
+
+    def min_fail(self, min="3"):
+        self._params.update(minFail=min)
+        return self
+
+    def pity(self, pity="5"):
+        self._params.update(pity=pity)
+        return self
+
+    def limit(self, limit="10"):
+        self._params.update(limit=limit)
+        return self
+
+    def sort(self, sort="asc"):
+        self._params.update(sort=sort)
+        return self
+
+    def sort_field(self, field="net_improvement"):
+        self._params.update(sortField=field)
+        return self
+
+    def period(self, period="towDay"):
+        self._params.update(period=period)
+        return self
+
+    def done(self):
+        return self._params
+
+
+class FilterBuilder():
+    # https://github.com/openshift/sippy/blob/master/pkg/api/README.md#filtering-and-sorting
+
+    def __init__(self):
+        self._params = {}
+        self._items = []
+
+    def filter(self, key, op, value, not_op=False):
+        '''
+          String operators are: contains, starts with, ends with, equals, is empty, is not empty.
+          Numerical operators are: =, !=, <, <=, >, >=
+          Array operators are: contains
+        '''
+
+        filter = {"columnField": key, "operatorValue": op, "value": value}
+        if not_op:
+            filter.update({"not": True})
+
+        self._items.append(filter)
+
+        return self
+
+    def link(self, AND=True):
+        if AND:
+            self._params.update(linkOperator="and")
+        else:
+            self._params.update(linkOperator="or")
+
+    def done(self):
+        if "linkOperator" not in self._params:
+            self.link()
+        self._params.update(items=self._items)
+        return json.dumps(self._params)
```

Comment:
@rioliu-rh I am not sure if it encodes the filter.
for example, for filter parameter `filter={"items":[{"columnField":"failed_test_names","operatorValue":"contains","value":"[sig-network][Feature:Router][apigroup:route.openshift.io] when FIPS is enabled the HAProxy router should not work when configured with a 1024-bit RSA key [Suite:openshift/conformance/parallel]"},{"columnField":"variants","not":true,"operatorValue":"contains","value":"never-stable"},{"columnField":"variants","not":true,"operatorValue":"contains","value":"aggregated"},{"columnField":"variants","operatorValue":"contains","value":"aws"},{"columnField":"variants","operatorValue":"contains","value":"amd64"},{"columnField":"variants","operatorValue":"contains","value":"ovn"},{"columnField":"variants","operatorValue":"contains","value":"upgrade"},{"columnField":"variants","operatorValue":"contains","value":"upgrade-minor"},{"columnField":"variants","operatorValue":"contains","value":"ha"}],"linkOperator":"and"}`, it should be `filter=%7B%22items%22%3A%5B%7B%22columnField%22%3A%22failed_test_names%22%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22%5Bsig-network%5D%5BFeature%3ARouter%5D%5Bapigroup%3Aroute.openshift.io%5D%20when%20FIPS%20is%20enabled%20the%20HAProxy%20router%20should%20not%20work%20when%20configured%20with%20a%201024-bit%20RSA%20key%20%5BSuite%3Aopenshift%2Fconformance%2Fparallel%5D%22%7D%2C%7B%22columnField%22%3A%22variants%22%2C%22not%22%3Atrue%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22never-stable%22%7D%2C%7B%22columnField%22%3A%22variants%22%2C%22not%22%3Atrue%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22aggregated%22%7D%2C%7B%22columnField%22%3A%22variants%22%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22aws%22%7D%2C%7B%22columnField%22%3A%22variants%22%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22amd64%22%7D%2C%7B%22columnField%22%3A%22variants%22%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22ovn%22%7D%2C%7B%22columnField%22%3A%22variants%22%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22upgrade%22%7D%2C%7B%22columnField%22%3A%22variants%22%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22upgrade-minor%22%7D%2C%7B%22columnField%22%3A%22variants%22%2C%22operatorValue%22%3A%22contains%22%2C%22value%22%3A%22ha%22%7D%5D%2C%22linkOperator%22%3A%22and%22%7D` in request.
it should be encoded.

anyway, if you could get response with filter, it is ok.

Reply:
yes the params can be encoded by request session

Reply:
ok, thanks